app/forms/otp_verification_form.rb

```diff
@@ -1,12 +1,25 @@
 class OtpVerificationForm
+  include ActiveModel::Model
+
+  validates :code, presence: true
+  validate :validate_code_matches_format
+  validate :validate_user_otp_presence
+  validate :validate_user_otp_expiration
+  validate :validate_code_equals_user_otp
+
   def initialize(user, code)
     @user = user
     @code = code
   end
 
   def submit
+    success = valid?
+
+    user.clear_direct_otp
+
     FormResponse.new(
-      success: valid_direct_otp_code?,
+      success: success,
+      errors: ErrorDetails.new(errors.details).flatten,
```

Comment:
This is rather unconventional, since `FormResponse` supports `errors` given as the `ActiveModel::Errors` object itself. However, we don't currently expose these errors directly to the user. We could generate some content for the error reasons, in which case this would result in logging both the full message as well as the `error_details` property of the `type` values, but that seemed like wasted effort since the messages wouldn't ever be shown to users, as well as wasted cost of logging storage for values which aren't strictly necessary?

I could be convinced to revise the approach to add the messages though, even if for consistency's sake.

Reply:
What if we added like a `error_details_only: true` option to FormResponse to add this behavior?

Reply:
I like that idea! I think it'd also be good to just omit `errors` and use the consistent `error_details` format for that option, so that it could be future-compatible to query if ever we decided to add human-readable error messages.

Reply:
yes definitely `error_details` only

Reply:
Added option in e02389b.